client/network/sync/src/blocks.rs

```diff
@@ -170,29 +172,51 @@ impl<B: BlockT> BlockCollection<B> {
 	}
 
 	/// Get a valid chain of blocks ordered in descending order and ready for importing into
-	/// blockchain.
-	pub fn drain(&mut self, from: NumberFor<B>) -> Vec<BlockData<B>> {
-		let mut drained = Vec::new();
-		let mut ranges = Vec::new();
+	/// the blockchain.
+	pub fn ready_blocks(&mut self, from: NumberFor<B>) -> Vec<BlockData<B>> {
+		let mut ready = Vec::new();
 
 		let mut prev = from;
 		for (start, range_data) in &mut self.blocks {
-			match range_data {
-				BlockRangeState::Complete(blocks) if *start <= prev => {
-					prev = *start + (blocks.len() as u32).into();
+			if *start > prev {
+				break
+			}
+			let len = match range_data {
+				BlockRangeState::Complete(blocks) => {
+					let len = (blocks.len() as u32).into();
+					prev = *start + len;
 					// Remove all elements from `blocks` and add them to `drained`
-					drained.append(blocks);
+					ready.append(blocks);
+					len
+				},
+				BlockRangeState::Queued { .. } => continue,
+				_ => break,
+			};
+			*range_data = BlockRangeState::Queued { len };
+		}
+
+		trace!(target: "sync", "{} blocks ready for import", ready.len());
+		ready
+	}
+
+	pub fn clear_queued(&mut self, from: NumberFor<B>) {
```

Comment:
The current implementation of this function could lead to deleting data in the queue that we don't yet have imported. You always set `prev = start + len`. If we have data that was downloaded after the last call to `ready_blocks`, we would also delete this.

I think we should change the `clear_queued` to take the hash of the first block that was returned by `ready_blocks`.

Internally we would add a `HashMap<Block::Hash, usize>` where the second parameter is the number of entries we need to delete in `self.blocks`. We would insert the value there in `ready_blocks`.

Reply:
Ah, that makes sense. Since we need to know the start number of the range along with the number of entries, instead of a `HashMap<Block::Hash, usize>` I went with a `HashMap<Block::Hash, (NumberFor<Block>, NumberFor<Block>)>`.